proposals/819-kubeflow-sdk/README.md

```diff
@@ -0,0 +1,422 @@
+data scientists and ML engineers to interact with various Kubeflow projects
+(Trainer, Katib, Spark Operator, Pipelines, Model Registry, KServe).
+
+## Motivation
+
+Kubeflow users come from diverse backgrounds, including cluster administrators, data scientists,
+and ML engineers (referred to as `ML User` in this proposal). While administrators are familiar
+with Kubernetes, ML Users typically are not. They prefer working in a familiar Python environment
+with popular data and ML frameworks like PyTorch, TensorFlow, and JAX for ML and Ray, Spark,
+Pandas, and Dask for data, while leveraging advanced hardware and cloud infrastructure
+to scale their experiments.
+
+Typically, ML Users want to:
+
+- Open JupyterLab.
+- Install required AI/ML libraries with `pip`.
+- Prototype ML code using PyTorch/TensorFlow APIs.
+- Launch jobs to create features from raw data and prepare training datasets.
+- Launch jobs to train, fine-tune, and optimize models across many GPUs.
+- Collect and analyze results within the same JupyterLab environment.
+
+One of the typical ML experience workflows can be seen at
+[this KubeCon + CloudNativeCon talk](https://youtu.be/Lgy4ir1AhYw?t=454).
+
+Currently, Kubeflow's ecosystem lacks simple interfaces for interacting with its components. This
+complexity hinders adoption among ML Users, as it requires Kubernetes expertise to use effectively.
+
+This document proposes the Kubeflow Python SDK, designed to streamline the experience of
+ML users when interacting with Kubeflow components. In the future, we may extend this SDK to
+support additional languages (e.g., Swift, Rust, Java) based on user demand.
+
+### Goals
+
+- Collaborate with ML Experience/IDE Working Group to include this SDK into their charter.
+
+- Introduce initial version of Kubeflow Python SDK that supports Kubeflow Trainer and Kubeflow
+  Katib APIs.
+
+  - **Note** that Kubeflow Katib will be probably renamed to Kubeflow Optimizer in the future.
+    Thus, in this proposal we will keep the Kubeflow Optimizer name moving forward.
+
+- Reduce duplication across `kubeflow-training` and `kubeflow-katib` SDKs by reusing code assets.
+
+- Design the SDK version control and release schedule.
+
+- Design synchronization scripts between Kubeflow control plane and SDK models.
+
+### No Goals
+
+- Consolidate Kubeflow Pipelines APIs into centralized Kubeflow SDK.
+
+  - As we discussed in the Google document, we want to postpone the `kfp` integration after
+    we release the initial version of Kubeflow SDK. The Kubeflow community will work towards
+    long-term goal to provide consistent experience for ML Users to interact with Kubeflow APIs.
+
+- Integrate Kubeflow Spark Operator, KServe, Model Registry into Kubeflow SDK.
+
+  - Similar to KFP, we will design the integration between those components in the future releases
+    of Kubeflow SDK.
+
+- Use the Kubeflow SDK to submit Kubernetes CRDs that are not part of Kubeflow projects,
```

Comment:
It seems here we are saying we are trying to abstract users from kubernetes as well. Can you add more clarifications on this? is it only for CRDs creations?

Reply:
It is as much as possible.

MLEs/Data Scientists shouldn't have to know about kubernetes to use kubeflow.

Reply:
This No Goal means that we are not planning to utilize this SDK to submit CRDs outside of Kubeflow Projects.
For example, leverage this SDK to create Argo Workflows, RayJobs, or KnativeServing.

Reply:
Actually Argo workflows are not a good example, since they are used by KFP, and  KnativeServing by Kserve.
RayJobs make sense.